### PR TITLE
Lightning upgrade

### PR DESCRIPTION
**Pull request by melo-gonzo**
This pr upgrades `pytorch_lightning` from 1.8.6 to 2.1.3. It also includes updates to all tests and most examples. The most notable changes are to account for the new `Embeddings` object returned from the model forward pass.

### CODE REVIEW

File: matsciml/datasets/__init__.py [hunk 1]
```diff
@@ -36,7 +36,8 @@
     "MultiDataset",
     "s2ef_devset",
     "is2redevset",
-    "carolinadb_devset" "nomad_devset",
```
Comment (laserkelvin) [inline matsciml/datasets/__init__.py:39]: How did we miss this?? LOL

File: matsciml/datasets/multi_dataset.py [hunk 2]
```diff
@@ -10,11 +10,13 @@
 from matsciml.common.registry import registry
 from matsciml.datasets.base import BaseLMDBDataset
 
+
 # quasi-registry of functions for collating based on dataset class name
-collate_registry = {
-    dset_name: dset_class.collate_fn
-    for dset_name, dset_class in registry.__entries__["datasets"].items()
-}
+def collate_registry() -> dict[str]:
```
Comment (laserkelvin) [inline matsciml/datasets/multi_dataset.py:15]: Can you explain why this changed?

Comment (melo-gonzo) [reply matsciml/datasets/multi_dataset.py:15]: Yeah - it is puzzling to me actually. I'm not sure if it's just limited to pytest, but the `collate registry` was being limited to just these datasets:
``` python
>>> collate_registry.keys()
>>> dict_keys(['CMDataset', 'M3GCMDataset', 'LiPSDataset', 'MaterialsProjectDataset', 'PyGMaterialsProjectDataset', 'PyGCdvaeDataset', 'CdvaeLMDBDataset', 'M3GMaterialsProjectDataset'])
```
Turning it into a function call brought back all the registered datasets. I'm guessing it has to do with how the imports are ordered in one of the files. The test what brought this issue up was `matsciml/datasets/tests/test_multi_dataset.py`.

File: matsciml/lightning/data_utils.py [hunk 3]
```diff
@@ -469,10 +469,12 @@ def persistent_workers(self, value: None | bool) -> None:
     def target_keys(self) -> dict[str, dict[str, list[str]]]:
         return self.datasets["train"].target_keys
 
+    # Cannot return None for dataloaders
+    # https://github.com/Lightning-AI/pytorch-lightning/issues/15703#issuecomment-1872664346
     def train_dataloader(self) -> DataLoader | None:
```
[the remark this answers is not in the document]
Comment (melo-gonzo) [reply matsciml/lightning/data_utils.py]: This is going to need to be `list | DataLoader`. Placing the dataloader into a list itself will mess up training - the batches will turn into lists then they cannot be iterated over as expected (expects dicts and uses batch.items()).

File: matsciml/models/dgl/m3gnet.py
[the remark this answers is not in the document]
Comment (melo-gonzo) [reply matsciml/models/dgl/m3gnet.py:1]: No problem - submitted PR #96 for this.

File: matsciml/datasets/multi_dataset.py (hunk 2 above)
Comment (laserkelvin) [reply matsciml/datasets/multi_dataset.py:15]: Could you add a quick comment to the function that explains that then? Otherwise I feel like we might go back later and forget why `collate_registry` is a function

File: matsciml/lightning/data_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (laserkelvin) [reply matsciml/lightning/data_utils.py]: Does `list` as an annotation work without a type inside it? If it does, I'm fine with that.

[the remark this answers is not in the document]
Comment (melo-gonzo) [reply matsciml/lightning/data_utils.py]: That's a good question - from what I can tell it does need something inside. I can put in list[Any] if that sounds good to you. Couldn't really find any info on how to annotate empty lists.

[the remark this answers is not in the document]
Comment (laserkelvin) [reply matsciml/lightning/data_utils.py]: Okay, just do `DataLoader | list[any]` as the annotation then, and it'll be good to go

[the remark this answers is not in the document]
Comment (melo-gonzo) [reply matsciml/lightning/data_utils.py]: Done ✅

File: matsciml/lightning/data_utils.py [hunk 4]
```diff
@@ -469,51 +469,60 @@ def persistent_workers(self, value: None | bool) -> None:
...
-            num_workers=self.hparams.num_workers,
-            collate_fn=data.collate_fn,
-            persistent_workers=self.persistent_workers,
-        )
+        if data:
+            loader = DataLoader(
+                data,
+                self.hparams.batch_size,
+                num_workers=self.hparams.num_workers,
+                shuffle=True,
```
Comment (laserkelvin) [inline matsciml/lightning/data_utils.py:510]: I don't think we should be shuffling any of the sets except for `train`

**Review by laserkelvin — CHANGES_REQUESTED** [review]

File: matsciml/lightning/data_utils.py (hunk 4 above)
Comment (melo-gonzo) [reply matsciml/lightning/data_utils.py]: oof, my bad on that one. corrected now

**Review by laserkelvin — APPROVED** [review]
LGTM, thanks for sitting through that!
